### PR TITLE
[addons] add addon filesystem callback functions to new style

### PR DESCRIPTION
**Pull request by AlwinEsch**
This is the next step to add addon callbacks to the new way.
After them are all parts who already in old libXBMC_addon.h included on
new interface headers.

As next comes the change of GUI callbacks.

## Types of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/addons/interfaces/Filesystem.cpp [hunk 1]
```diff
@@ -0,0 +1,582 @@
...
+  return static_cast<CFile*>(file)->Seek(position, whence);
+}
+
+int Interface_Filesystem::truncate_file(void* kodiBase, void* file, int64_t size)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  if (addon == nullptr || file == nullptr)
+  {
+    CLog::Log(LOGERROR, "Interface_Filesystem::%s - invalid data (addon='%p', file='%p')", __FUNCTION__, addon, file);
+    return 0.0f;
```
Comment (ksooo) [inline xbmc/addons/interfaces/Filesystem.cpp]: This function returns an int. So this should read <code>return 0;</code> or even better <code>return -1;</code> as <code>CFile::Truncate</code> returns -1 in error case.

File: xbmc/addons/interfaces/Filesystem.cpp [hunk 2]
```diff
@@ -0,0 +1,582 @@
...
+  static_cast<CFile*>(file)->Flush();
+}
+
+int64_t Interface_Filesystem::seek_file(void* kodiBase, void* file, int64_t position, int whence)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  if (addon == nullptr || file == nullptr)
+  {
+    CLog::Log(LOGERROR, "Interface_Filesystem::%s - invalid data (addon='%p', file='%p')", __FUNCTION__, addon, file);
+    return 0.0f;
```
Comment (ksooo) [inline xbmc/addons/interfaces/Filesystem.cpp]: This function returns an int64_t. So this should read <code>return 0;</code> or even better <code>return -1;</code> as <code>CFile::Seek</code> returns -1 in error case.

File: xbmc/addons/interfaces/Filesystem.cpp [hunk 3]
```diff
@@ -0,0 +1,582 @@
...
+  return static_cast<CFile*>(file)->Truncate(size);
+}
+
+int64_t Interface_Filesystem::get_file_position(void* kodiBase, void* file)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  if (addon == nullptr || file == nullptr)
+  {
+    CLog::Log(LOGERROR, "Interface_Filesystem::%s - invalid data (addon='%p', file='%p')", __FUNCTION__, addon, file);
+    return 0;
```
Comment (ksooo) [inline xbmc/addons/interfaces/Filesystem.cpp]: <code>return -1;</code>?

File: xbmc/addons/interfaces/Filesystem.cpp [hunk 4]
```diff
@@ -0,0 +1,582 @@
...
+  return static_cast<CFile*>(file)->GetPosition();
+}
+
+int64_t Interface_Filesystem::get_file_length(void* kodiBase, void* file)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  if (addon == nullptr || file == nullptr)
+  {
+    CLog::Log(LOGERROR, "Interface_Filesystem::%s - invalid data (addon='%p', file='%p')", __FUNCTION__, addon, file);
+    return 0;
```
Comment (ksooo) [inline xbmc/addons/interfaces/Filesystem.cpp]: <code>return -1;</code>?

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/Filesystem.h [hunk 5]
```diff
@@ -0,0 +1,1486 @@
...
+    /// @return map with all present properties
+    ///
+    const std::map<std::string, std::string>& GetProperties() const { return m_property; }
+    //----------------------------------------------------------------------------
+
+  private:
+    std::string m_label;
+    std::string m_title;
+    std::string m_path;
+    std::map<std::string, std::string> m_property;
```
Comment (ksooo) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/Filesystem.h]: should either read m_propertyMap or m_properties

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/Filesystem.h [hunk 6]
```diff
@@ -0,0 +1,1486 @@
...
+  //============================================================================
+  ///
+  /// @ingroup cpp_kodi_vfs
+  /// @brief Return a size aligned to the chunk size at least as large as the chunk size.
+  ///
+  /// @param[in] chunk The chunk size
+  /// @param[in] minimum The minimum size (or maybe the minimum number of chunks?)
+  /// @return The aligned size
+  ///
+  static inline unsigned int GetChunkSize(unsigned int chunk, unsigned int minimum)
```
Comment (ksooo) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/Filesystem.h]: Some of the global function in the header are static, others are not. What is the criteria? Should be consistent.

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/Filesystem.h [hunk 7]
```diff
@@ -0,0 +1,1486 @@
...
+    /// @return                 number of successfully read bytes if any bytes
+    ///                         were read and stored in buffer, zero if no bytes
+    ///                         are available to read (end of file was reached)
+    ///                         or undetectable error occur, -1 in case of any
+    ///                         explicit error
+    ///
+    ssize_t Read(void* ptr, size_t size)
+    {
+      if (!m_file)
+        return 0;
```
Comment (ksooo) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/Filesystem.h]: <code>return -1;</code>?

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/Filesystem.h [hunk 8]
```diff
@@ -0,0 +1,1486 @@
...
+    /// @param[in] size         Size of the data to write.
+    /// @return                 number of successfully written bytes if any
+    ///                         bytes were written, zero if no bytes were
+    ///                         written and no detectable error occur,-1 in case
+    ///                         of any explicit error
+    ///
+    ssize_t Write(const void* ptr, size_t size)
+    {
+      if (!m_file)
+        return 0;
```
Comment (ksooo) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/Filesystem.h]: <code>return -1;</code>?

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/Filesystem.h [hunk 9]
```diff
@@ -0,0 +1,1486 @@
...
+    ///
+    /// @return                         Returns the resulting offset location as
+    ///                                 measured in bytes from the beginning of
+    ///                                 the file. On error, the value -1 is
+    ///                                 returned.
+    ///
+    int64_t Seek(int64_t position, int whence = SEEK_SET)
+    {
+      if (!m_file)
+        return 0;
```
Comment (ksooo) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/Filesystem.h]: <code>return -1;</code>?

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/Filesystem.h [hunk 10]
```diff
@@ -0,0 +1,1486 @@
...
+    ///
+    /// @ingroup cpp_kodi_vfs_CFile
+    /// @brief Get the file size of an open file.
+    ///
+    /// @return                 The requested size.
+    ///
+    int64_t GetLength()
+    {
+      if (!m_file)
+        return 0;
```
Comment (ksooo) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/Filesystem.h]: <code>return -1;</code>?

**Review by ksooo — CHANGES_REQUESTED** [review]
Some suggestions and questions.

File: xbmc/addons/interfaces/Filesystem.cpp [hunk 11]
```diff
@@ -0,0 +1,592 @@
...
+}
+
+static void CFileItemListToVFSDirEntries(VFSDirEntry* entries,
+                                         unsigned int num_entries,
+                                         const CFileItemList& items)
+{
+  if (!entries)
+    return;
+
+  int toCopy = std::min(num_entries, static_cast<unsigned int>(items.Size()));
```
Comment (AlwinEsch) [inline xbmc/addons/interfaces/Filesystem.cpp]: Becomes changed, was a bit a c&p https://github.com/xbmc/xbmc/blob/master/xbmc/addons/interfaces/Addon/AddonCallbacksAddon.cpp#L582-L598 :)

Comment (ksooo) [inline xbmc/addons/interfaces/Filesystem.cpp]: Okay.

**Comment by AlwinEsch** [comment]
Is updated, thanks for review!

**Review by ksooo — APPROVED** [review]

**Comment by ksooo** [comment]
Okay to merge, imo.

**Comment by AlwinEsch** [comment]
jenkins build this please

**Pull request merged**